### ISSUE TITLE
Annotation to attach detached mock

### ISSUE THREAD

**Issue by musketyr** [spockframework/spock#779]
Attaching mock manually is not very practical. It would be nice if Spock has an annotation to attach detached mocks. Here's working implementation I'm using locally:

```
import org.spockframework.runtime.extension.ExtensionAnnotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.FIELD)
@ExtensionAnnotation(DetachedExtension.class)
public @interface Detached { }
```

```
import org.spockframework.mock.MockUtil;
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension;
import org.spockframework.runtime.extension.IMethodInterceptor;
import org.spockframework.runtime.extension.IMethodInvocation;
import org.spockframework.runtime.model.FieldInfo;
import spock.lang.Specification;

public class DetachedExtension extends AbstractAnnotationDrivenExtension<Detached> {

    private static final MockUtil MOCK_UTIL = new MockUtil();

    @Override
    public void visitFieldAnnotation(Detached annotation, FieldInfo field) {
        field.getParent().addSetupInterceptor(new IMethodInterceptor() {
            @Override
            public void intercept(IMethodInvocation invocation) throws Throwable {
                MOCK_UTIL.attachMock(field.readValue(invocation.getInstance()), (Specification) invocation.getInstance());
                invocation.proceed();
            }
        });
        field.getParent().addCleanupInterceptor(new IMethodInterceptor() {
            @Override
            public void intercept(IMethodInvocation invocation) throws Throwable {
                MOCK_UTIL.detachMock(field.readValue(invocation.getInstance()));
                invocation.proceed();
            }
        });
    }
}
```

Which can be used for example within Grails Unit Tests using AutowiredTest annotation:

```
@Autowired @Detached FooService fooService
```

**Comment by leonard84** [comment]
Spocks `org.spockframework.spring.SpringMockTestExecutionListener`automatically attaches detached mocks it should be automatically registered by spocks  `SpringExtension`. Did you include the `spock-spring` module?

**Comment by musketyr** [comment]
Grails unit tests are not recognized as Spring tests but I guess there might be other use cases (e.g TCK also providing mocks for implementations).

see https://medium.com/@musketyr/leveraging-spock-spring-module-in-grails-unit-tests-e839340de272 for context.

**Comment by musketyr** [comment]
Generally speaking as `DetachedMockFactory` is part of the Spock core library it should be possible to use it in a simple way even without Spring module.

**Comment by leonard84** [comment]
Implemented with #790 the annotation is called `@AutoAttach`, thanks for your input.
